compiler/types/Multiplicity.hs

```diff
@@ -42,6 +42,219 @@ import {-# SOURCE #-} Type( eqType, splitTyConApp_maybe, mkTyConApp )
 import PrelNames (multMulTyConKey)
 import Unique (hasKey)
 
+{-
+Note [Linear types]
+~~~~~~~~~~~~~~~~~~~
+This module is the entry point for linear types.
+
+The detailed design is in the _Linear Haskell_ article
+[https://arxiv.org/abs/1710.09756]. Other important resources in the linear
+types implementation wiki page
+[https://gitlab.haskell.org/ghc/ghc/wikis/linear-types/implementation], and the
+proposal [https://github.com/ghc-proposals/ghc-proposals/pull/111] which
+describes the concrete design at length.
+
+For the busy developer, though, here is a high-level view of linear types is the following:
+
+- Function arrows are annotated with a multiplicity (as defined by type `Mult`
+  and its smart constructors in this module)
+- Multiplicities can be reified in Haskell as types of kind
+  `GHC.Types.Multiplicity`
+- Ground (that is, without variable) can be `One` or `Many` (`Many` is generally
+  rendered as ω in the scientific literature). Functions whose type is annotated
+  with `One` are linear functions, functions whose type is annotated with
+  `Omega` are regular functions, often called “unrestricted” to contrast them
+  with linear functions.
+- A linear function is defined as a function such that *if* its result is
+  consumed exactly once, *then* its argument is consumed exactly once. You can
+  think of “consuming exactly once” as evaluating a value in normal form exactly
+  once (though not necessarily in one go). The _Linear Haskell_ article (see
+  infra) has a more precise definition of “consuming exactly once”.
+- Data types can have unrestricted fields (the canonical example being the
+  `Unrestricted` data type), then these don't need to be consumed for a value to
+  be consumed exactly once. So consuming a value of type `Unrestricted` exactly
+  once means forcing it at least once.
```

Comment:
I don't understand what is being forced at least once. The value of type Unrestricted needs to be under case-1 exactly once, the field can be forced arbitrarily many times, what has to be forced at least once?

Reply:
@aspiwack I haven't fixed this one as I'm not sure; leaving to you.